components/support/nimbus-fml/src/editing/error_converter.rs

```diff
@@ -0,0 +1,121 @@
+        &self,
+        feature_def: &FeatureDef,
+        feature_value: &Value,
+        error: FeatureValidationError,
+    ) -> FMLError {
+        let values = ValuesFinder::new(self.enum_defs, feature_def, feature_value);
+        let long_message = self.long_message(&values, &error);
+        FMLError::ValidationError(error.path.path, long_message)
+    }
+
+    #[allow(dead_code)]
+    #[cfg(feature = "client-lib")]
+    pub(crate) fn convert_into_editor_errors(
+        &self,
+        feature_def: &FeatureDef,
+        feature_value: &Value,
+        src: &str,
+        errors: &Vec<FeatureValidationError>,
+    ) -> Vec<FmlEditorError> {
+        let mut editor_errors: Vec<_> = Default::default();
+        let values = ValuesFinder::new(self.enum_defs, feature_def, feature_value);
+        for e in errors {
+            let message = self.long_message(&values, e);
+            let highlight = e.path.last_token().map(str::to_string);
+            let (line, col) = e.path.line_col(src);
+            let error = FmlEditorError {
+                message,
+                line: line as u32,
+                col: col as u32,
+                highlight,
+            };
+            editor_errors.push(error);
+        }
+        editor_errors
+    }
+
+    pub(crate) fn convert_object_error(&self, error: FeatureValidationError) -> FMLError {
+        let message = self.message(&error);
+        FMLError::ValidationError(error.path.path, message)
+    }
+}
+
+impl ErrorConverter<'_> {
+    fn long_message(&self, values: &ValuesFinder, error: &FeatureValidationError) -> String {
+        let message = self.message(error);
+        let suggestions = self.suggested_replacements(error, values);
+        let dym = did_you_mean(suggestions);
+        format!("{message}{dym}")
+    }
+
+    fn message(&self, error: &FeatureValidationError) -> String {
+        let token = error.path.last_token().unwrap_or("unknown");
+        error.kind.message(token)
+    }
+
+    fn suggested_replacements(
+        &self,
+        error: &FeatureValidationError,
+        values: &ValuesFinder,
+    ) -> BTreeSet<String> {
+        match &error.kind {
+            ErrorKind::InvalidKey(t, _)
+            | ErrorKind::InvalidValue(t)
+            | ErrorKind::TypeMismatch(t) => values.all(t),
+            ErrorKind::InvalidPropKey(valid, _) => valid.to_owned(),
+            ErrorKind::InvalidNestedValue(_, _) => Default::default(),
+        }
+    }
+}
+
+pub(crate) fn did_you_mean(words: BTreeSet<String>) -> String {
+    if words.is_empty() {
+        "".to_string()
+    } else if words.len() == 1 {
+        format!("; did you mean \"{}\"?", words.iter().next().unwrap())
+    } else if words.len() == 2 {
+        format!(
+            "; did you mean \"{}\" or \"{}\"?",
+            words.first().unwrap(),
+            words.last().unwrap()
+        )
+    } else {
+        let mut words = words.into_iter().collect::<Vec<_>>();
```

Comment:
This would be a great place to use [`Itertools::join`](https://docs.rs/itertools/latest/itertools/fn.join.html) from the itertools crate.